File: src/test/java/com/joecollins/graphics/components/ListingFrameBuilderTest.java
package com.joecollins.graphics.components;

import static org.junit.Assert.assertEquals;

import com.joecollins.bindings.BindableList;
import com.joecollins.bindings.Binding;
import com.joecollins.graphics.utils.BindableWrapper;
import java.awt.Color;
import java.util.ArrayList;
import java.util.List;
import org.apache.commons.lang3.tuple.ImmutableTriple;
import org.apache.commons.lang3.tuple.Triple;
import org.junit.Test;

public class ListingFrameBuilderTest {

  @Test
  public void testBasicListingFrame() {
    BindableList<Triple<String, Color, String>> list = new BindableList<>();
    list.add(ImmutableTriple.of("JUSTIN TRUDEAU", Color.RED, "LIBERAL"));
    list.add(ImmutableTriple.of("ANDREW SCHEER", Color.BLUE, "CONSERVATIVE"));

    BarFrame frame =
        ListingFrameBuilder.of(list, Triple::getLeft, Triple::getRight, Triple::getMiddle)
            .withHeader(Binding.fixedBinding("HEADER"))
            .withSubhead(Binding.fixedBinding("SUBHEAD"))
            .build();
    assertEquals(0, frame.getNumLines());
    assertEquals(2, frame.getNumBars());
    assertEquals(0, frame.getMin().doubleValue(), 1e-6);
    assertEquals(1, frame.getMax().doubleValue(), 1e-6);
    assertEquals("HEADER", frame.getHeader());
    assertEquals("SUBHEAD", frame.getSubheadText());

    assertEquals("JUSTIN TRUDEAU", frame.getLeftText(0));
    assertEquals("LIBERAL", frame.getRightText(0));
    assertEquals(1, frame.getSeries(0).size());
    assertEquals(Color.RED, frame.getSeries(0).get(0).getLeft());
    assertEquals(1.0, frame.getSeries(0).get(0).getRight().doubleValue(), 1e-6);

    assertEquals("ANDREW SCHEER", frame.getLeftText(1));
    assertEquals("CONSERVATIVE", frame.getRightText(1));
    assertEquals(1, frame.getSeries(1).size());
    assertEquals(Color.BLUE, frame.getSeries(1).get(0).getLeft());
    assertEquals(1.0, frame.getSeries(1).get(0).getRight().doubleValue(), 1e-6);
  }

  @Test
  public void testBasicListingFrameWithListBinding() {
    List<Triple<String, Color, String>> list = new ArrayList<>();
    list.add(ImmutableTriple.of("JUSTIN TRUDEAU", Color.RED, "LIBERAL"));
    list.add(ImmutableTriple.of("ANDREW SCHEER", Color.BLUE, "CONSERVATIVE"));

    BarFrame frame =
        ListingFrameBuilder.of(() -> list, Triple::getLeft, Triple::getRight, Triple::getMiddle)
            .withHeader(Binding.fixedBinding("HEADER"))
            .withSubhead(Binding.fixedBinding("SUBHEAD"))
            .build();
    assertEquals(0, frame.getNumLines());
    assertEquals(2, frame.getNumBars());
    assertEquals(0, frame.getMin().doubleValue(), 1e-6);
    assertEquals(1, frame.getMax().doubleValue(), 1e-6);
    assertEquals("HEADER", frame.getHeader());
    assertEquals("SUBHEAD", frame.getSubheadText());

    assertEquals("JUSTIN TRUDEAU", frame.getLeftText(0));
    assertEquals("LIBERAL", frame.getRightText(0));
    assertEquals(1, frame.getSeries(0).size());
    assertEquals(Color.RED, frame.getSeries(0).get(0).getLeft());
    assertEquals(1.0, frame.getSeries(0).get(0).getRight().doubleValue(), 1e-6);

    assertEquals("ANDREW SCHEER", frame.getLeftText(1));
    assertEquals("CONSERVATIVE", frame.getRightText(1));
    assertEquals(1, frame.getSeries(1).size());
    assertEquals(Color.BLUE, frame.getSeries(1).get(0).getLeft());
    assertEquals(1.0, frame.getSeries(1).get(0).getRight().doubleValue(), 1e-6);
  }

  @Test
  public void testBasicFixedListFrame() {
    List<Triple<BindableWrapper<String>, BindableWrapper<Color>, BindableWrapper<String>>> list =
        new ArrayList<>();
    list.add(
        ImmutableTriple.of(
            new BindableWrapper<>("JUSTIN TRUDEAU"),
            new BindableWrapper<>(Color.RED),
            new BindableWrapper<>("LIBERAL")));
    list.add(
        ImmutableTriple.of(
            new BindableWrapper<>("ANDREW SCHEER"),
            new BindableWrapper<>(Color.BLUE),
            new BindableWrapper<>("CONSERVATIVE")));

    BarFrame frame =
        ListingFrameBuilder.ofFixedList(
                list,
                e -> e.getLeft().getBinding(),
                e -> e.getRight().getBinding(),
                e -> e.getMiddle().getBinding())
            .withHeader(Binding.fixedBinding("HEADER"))
            .withSubhead(Binding.fixedBinding("SUBHEAD"))
            .build();
    assertEquals(0, frame.getNumLines());
    assertEquals(2, frame.getNumBars());
    assertEquals(0, frame.getMin().doubleValue(), 1e-6);
    assertEquals(1, frame.getMax().doubleValue(), 1e-6);
    assertEquals("HEADER", frame.getHeader());
    assertEquals("SUBHEAD", frame.getSubheadText());

    assertEquals("JUSTIN TRUDEAU", frame.getLeftText(0));
    assertEquals("LIBERAL", frame.getRightText(0));
    assertEquals(1, frame.getSeries(0).size());
    assertEquals(Color.RED, frame.getSeries(0).get(0).getLeft());
    assertEquals(1.0, frame.getSeries(0).get(0).getRight().doubleValue(), 1e-6);

    assertEquals("ANDREW SCHEER", frame.getLeftText(1));
    assertEquals("CONSERVATIVE", frame.getRightText(1));
    assertEquals(1, frame.getSeries(1).size());
    assertEquals(Color.BLUE, frame.getSeries(1).get(0).getLeft());
    assertEquals(1.0, frame.getSeries(1).get(0).getRight().doubleValue(), 1e-6);
  }
}

File: src/test/kotlin/com/joecollins/graphics/components/ListingFrameBuilderTest.kt
package com.joecollins.graphics.components

import com.joecollins.graphics.components.ListingFrameBuilder.Companion.of
import com.joecollins.graphics.components.ListingFrameBuilder.Companion.ofFixedList
import com.joecollins.pubsub.Publisher
import com.joecollins.pubsub.asOneTimePublisher
import org.awaitility.Awaitility
import org.hamcrest.core.IsEqual
import org.junit.Assert
import org.junit.Test
import java.awt.Color
import java.util.concurrent.TimeUnit

class ListingFrameBuilderTest {
    @Test
    fun testBasicListingFrameWithListBinding() {
        val list = listOf(
            Triple("JUSTIN TRUDEAU", Color.RED, "LIBERAL"),
            Triple("ANDREW SCHEER", Color.BLUE, "CONSERVATIVE")
        )
        val frame: BarFrame = of(list.asOneTimePublisher(), { it.first }, { it.third }) { it.second }
            .withHeader("HEADER".asOneTimePublisher())
            .withSubhead("SUBHEAD".asOneTimePublisher())
            .build()
        Assert.assertEquals(0, frame.numLines.toLong())
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.numBars }, IsEqual(2))
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.max.toDouble() }, IsEqual(1.0))
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.min.toDouble() }, IsEqual(0.0))
        Assert.assertEquals("HEADER", frame.header)
        Assert.assertEquals("SUBHEAD", frame.subheadText)
        Assert.assertEquals("JUSTIN TRUDEAU", frame.getLeftText(0))
        Assert.assertEquals("LIBERAL", frame.getRightText(0))
        Assert.assertEquals(1, frame.getSeries(0).size.toLong())
        Assert.assertEquals(Color.RED, frame.getSeries(0)[0].first)
        Assert.assertEquals(1.0, frame.getSeries(0)[0].second.toDouble(), 1e-6)
        Assert.assertEquals("ANDREW SCHEER", frame.getLeftText(1))
        Assert.assertEquals("CONSERVATIVE", frame.getRightText(1))
        Assert.assertEquals(1, frame.getSeries(1).size.toLong())
        Assert.assertEquals(Color.BLUE, frame.getSeries(1)[0].first)
        Assert.assertEquals(1.0, frame.getSeries(1)[0].second.toDouble(), 1e-6)
    }

    @Test
    fun testBasicFixedListFrame() {
        val list = listOf(
            Triple(
                Publisher("JUSTIN TRUDEAU"),
                Publisher(Color.RED),
                Publisher("LIBERAL")
            ),
            Triple(
                Publisher("ANDREW SCHEER"),
                Publisher(Color.BLUE),
                Publisher("CONSERVATIVE")
            )
        )
        val frame = ofFixedList(list, { it.first }, { it.third }) { it.second }
            .withHeader("HEADER".asOneTimePublisher())
            .withSubhead("SUBHEAD".asOneTimePublisher())
            .build()
        Assert.assertEquals(0, frame.numLines.toLong())
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.numBars }, IsEqual(2))
        Assert.assertEquals(0.0, frame.min.toDouble(), 1e-6)
        Assert.assertEquals(1.0, frame.max.toDouble(), 1e-6)
        Assert.assertEquals("HEADER", frame.header)
        Assert.assertEquals("SUBHEAD", frame.subheadText)
        Assert.assertEquals("JUSTIN TRUDEAU", frame.getLeftText(0))
        Assert.assertEquals("LIBERAL", frame.getRightText(0))
        Assert.assertEquals(1, frame.getSeries(0).size.toLong())
        Assert.assertEquals(Color.RED, frame.getSeries(0)[0].first)
        Assert.assertEquals(1.0, frame.getSeries(0)[0].second.toDouble(), 1e-6)
        Assert.assertEquals("ANDREW SCHEER", frame.getLeftText(1))
        Assert.assertEquals("CONSERVATIVE", frame.getRightText(1))
        Assert.assertEquals(1, frame.getSeries(1).size.toLong())
        Assert.assertEquals(Color.BLUE, frame.getSeries(1)[0].first)
        Assert.assertEquals(1.0, frame.getSeries(1)[0].second.toDouble(), 1e-6)
    }
}
